### DISCUSSION TITLE
Clone Tenant

### DISCUSSION THREAD

**Discussion by TimTracy** [OrchardCMS/OrchardCore#11936] [Q&A]
After setting up several websites, and going through the process of enabling features and other customizations for each tenant it would be nice to be able to clone a Startup tenant with all this work already done. I do not see a clone option in the SaaS area.  The tables will need to be duplicated  and the folder at OrchardCore.Cms.Web\App_Data\Sites.  Any one else trying something like this?
Thanks.

Sorry for the delay, but off a few days for the 4th.
Spent hours scripting Orchard Core tables and inserting records from Startup site and working with App_Data folder without much luck of Cloning Web Site successfully..  I would like to clone a Orchard Core Startup Web Site, not a live Web Site being used by a client.  I spend days adding features, templates, etc. to the Blog theme for a new Website.  I used the following technique to clone a existing Web Site in Orchard Core. This could be Setup to execute from a front end.
1. From Saas site create your New Website.
2. Using SQL statements truncate records in the 15 Orchard Core tables for the new WebSite you created. The  Document table was done manually due to table constraits.
3. Using SQl statements insert records into the new WebSite tables from the StartUp Website tables.
4. Test new WebSite for features from the Startup website.

Once I got the SQL statements working I can clone a New Website in a couple minutes  and it appears as a tenant in the Saas website.
This will save me many days of working bringing up a New Website and adding many features, templates, etc.    
 
Thanks,

**Comment by hishamco** [comment]
Cloning might be a good option but it needs more thinking because by default it will use the same database - which is not the case in almost the scenarios - also it needs to know in advance what's the table prefix that will be used

Let us hear the team and community feedback

**Comment by Skrypt** [comment]
Cloning might require also to copy all the media from one Azure Blob Storage account to another.

**Reply by hishamco** [reply]
That's another issue, so IMHO we should leave this as custom feature, to reduce the possiblities of the required changes in the system

**Reply by TimTracy** [reply]
Clone Website would not be a front end option, but could be.

using SQLServer,  Prompt for name of new website in storedprocedures, this would be used as table extension also for new WebSite.
Check make sure new website name has not been used.

1. Store Procedure would create new table from the 15 tables used by a OrchardCore WebSite with the new Website name.
	SQL would be something like:
        CREATE  'WebSite1_AliasPartIndex' FROM  'StartUp1_AliasPartIndex';
	CREATE ' WebSite1_AutoroutePartIndex' FROM  'Startup1_AutoroutePartIndex';
	etc ...	

2. Copy folder OrchardCore.Cms.Web\App_Data\Sites\StartUpSite folder to OrchardCore.Cms.Web\App_Data\Sites\WebSite1
    update the appsettings JSON file with the new Website name and sql connection login info.
	"ConnectionString": "Server=tcp:",
  	"DatabaseProvider": "SqlConnection",
  	"Description": "WebSite",
 	 "RecipeName": "Blog",
  	"Secret": "c977d6d9-xxxxxx-xxxxx-a38e77a67c9e",
  	"TablePrefix": "WebSite1"

3. Not sure at this point the new tenant will show on the SaaS website.
 	The 7 tables used by a OrchardCore Saas website do not contain the Websites already setup.
	The list of WebSites setup used by the SaaS WebSite must get the list from?
		OrchardCore.Cms.Web\App_Data\Sites\

			OrchardCore.Cms.Web\App_Data\Sites\StartUpSite1
			OrchardCore.Cms.Web\App_Data\Sites\WebSite

I may give this a try.
Thanks for your input.

**Reply by hishamco** [reply]
I'm glad to accept your changed in Orchard Core Contribute (OCC) if there's no plan to implement in OC, but again it needs a lot of thinking because there are many things we should consider

Thanks

**Comment by MikeAlhayek** [comment]
Another issue with cloning a tenant is data protection. Each tenant has different key to protect data. If you were to clone data, you'll have to unprotect all data using old tenant's key and re-protect the same data with the new key.

The question is why would you want to clone a tenant's data (users, password, media, documents.....)? If you just want one tenant to look like another, shouldn't you use recipe to export what you need into a new tenant instead?

**Reply by ns8482e** [reply]
For cloning tenant Missing piece of puzzle  in recipe is that, you run recipe in context of current existing tenant.

whereas clone tenant should create the tenant and run recipe within new tenant’s scope

**Comment by ns8482e** [comment]
IMO clone tenant is deployment plan that should export tenant settings and content to a recipe/package with some configuration variables, that the  inputs  are supplied during import
